plasmapy/diagnostics/charged_particle_radiography.py

```diff
@@ -1115,118 +1263,161 @@ def max_deflection(self):
 
-        # Generate the histogram
-        intensity, h, v = np.histogram2d(
-            xloc, yloc, range=size.to(u.m).value, bins=bins
+        If the `Intensity` histogram contains zeros, `OD` will contain
+        -`~numpy.inf` values. These can be easily replaced with zeros
+        if desired for plotting using `~numpy.nan_to_num`
+
+        ``OD = np.nan_to_num(OD, neginf=0)``
+
+    Returns
+    -------
+    hax : `~astropy.units.Quantity` array shape ``(hbins,)``
+        The horizontal axis of the synthetic radiograph in meters.
+
+    vax : `~astropy.units.Quantity` array shape ``(vbins, )``
+        The vertical axis of the synthetic radiograph in meters.
+
+    intensity : `~numpy.ndarray`, shape ``(hbins, vbins)``
+        The number of particles counted in each bin of the histogram.
+    """
+
+    # results_dict will raise an error if the simulation has not been run.
+    if isinstance(obj, Tracker):
+        d = obj.results_dict
+
+    # Check if dictionary-like (dict or npz file)
+    elif isinstance(obj, dict):
+        d = obj
+    else:
+        raise ValueError(
+            "The first argument of synthetic_radiograph must be "
+            "either a cpr.Tracker or an "
+            "output dictionary from cpr.Tracker"
         )
 
-        # h, v are the bin edges: compute the centers to produce arrays
-        # of the right length (then trim off the extra point)
-        h = ((h + np.roll(h, -1)) / 2)[0:-1]
-        v = ((v + np.roll(v, -1)) / 2)[0:-1]
+    if bins is None:
+        bins = [200, 200]
 
-        # Throw a warning if < 50% of the particles are included on the
-        # histogram
-        percentage = np.sum(intensity) / self.nparticles
-        if percentage < 0.5:
-            warnings.warn(
-                f"Only {percentage:.2%} of the particles are shown "
-                "on this synthetic radiograph. Consider increasing "
-                "the size to include more.",
-                RuntimeWarning,
-            )
+    # Note that, at the end of the simulation, all particles were moved
+    # into the image plane.
+
+    # If ignore_grid is True, use the predicted positions in the absence of
+    # simulated fields
+    if ignore_grid:
+        xloc = d["x0"]
+        yloc = d["y0"]
+    else:
+        xloc = d["x"]
+        yloc = d["y"]
+
+    if size is None:
+        # If a detector size is not given, choose a size based on the
+        # particle positions
+        w = np.max([np.max(np.abs(xloc)), np.max(np.abs(yloc))])
+        size = np.array([[-w, w], [-w, w]]) * u.m
+    elif not isinstance(size, u.Quantity) or not size.unit.is_equivalent(u.m):
+        raise TypeError(
+            "``size`` must be an `~astropy.units.Quantity` "
+            "object with units convertable to meters."
+        )
+    elif size.shape != (2, 2):
+        raise ValueError(
+            "``size`` must have shape ``(2,2)`` corresponding to "
+            "``[[xmin, xmax], [ymin, ymax]]``."
+        )
```

Comment:
```suggestion
        elif not isinstance(size, u.Quantity):
        raise TypeError(
            "Argument `size` must be an astropy.units.Quantity object with "
            "units convertable to meters."
        )
    elif not size.unit.is_equivalent(u.m):
        raise ValueError(
            "Argument `size` must have units convertible to meters."
        )
    elif size.shape != (2, 2):
        raise ValueError(
            f"Argument `size` must have shape (2, 2), but got {size.shape}."
        )
```

I broke this up because the not a `Quantity` is a `TypeError` but the rest are `ValueError`'s.

Note that all the back-tick and tilde syntax used in docstrings don't work for warnings or issues.  The syntax is specific to `sphinx` stuff.